### PR TITLE
Remove arch_opt enum and use cpu features flags instead

### PR DESCRIPTION
**Pull request by alonre24**
**Describe the changes in the pull request**

As we move to use more hardware flags with various configurations, we remove the enum that assumes a relation of order between these flags, and check the features flags explicitly instead, whenever they are required. 

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/spaces/IP_space.cpp [hunk 1]
```diff
@@ -30,36 +30,31 @@ dist_func_t<float> IP_FP32_GetDistFunc(size_t dim, const Arch_Optimization arch_
         return ret_dist_func;
     }
 #ifdef CPU_FEATURES_ARCH_X86_64
-
-    switch (arch_opt) {
-    case ARCH_OPT_AVX512_BW_VBMI2:
-    case ARCH_OPT_AVX512_F:
+    auto features = arch_opt.features;
```
Comment (meiravgri) [inline src/VecSim/spaces/IP_space.cpp:33]: if we return in every if block, the pre declaration of `ret_dist_func` is redundant

File: tests/unit/test_spaces.cpp [hunk 2]
```diff
@@ -14,6 +14,12 @@
 #include "VecSim/types/bfloat16.h"
 #include "VecSim/spaces/IP_space.h"
 #include "VecSim/spaces/L2_space.h"
+#include "VecSim/spaces/functions/AVX512.h"
+#include "VecSim/spaces/functions/AVX.h"
+#include "VecSim/spaces/functions/AVX512BW_VBMI2.h"
+#include "VecSim/spaces/functions/AVX2.h"
+#include "VecSim/spaces/functions/SSE3.h"
+#include "VecSim/spaces/functions/SSE.h"
```
Comment (meiravgri) [inline tests/unit/test_spaces.cpp:22]: are those still necessary?

File: tests/unit/test_spaces.cpp [hunk 3]
```diff
@@ -115,25 +135,21 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
...
-    case ARCH_OPT_AVX512_BW_VBMI2:
-    case ARCH_OPT_AVX512_F:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F);
+    if (optimization.features.avx512f) {
+        arch_opt_func = L2_FP32_GetDistFunc(dim, optimization);
         ASSERT_EQ(baseline, arch_opt_func(v, v2, dim)) << "AVX512 with dim " << dim;
-    case ARCH_OPT_AVX2:
-    case ARCH_OPT_AVX:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX);
+        // Unet avx512f flag, so we'll choose the next optimization (AVX).
```
Comment (meiravgri) [inline tests/unit/test_spaces.cpp:141]:
```suggestion
        // Unset avx512f flag, so we'll choose the next optimization (AVX).
```

File: tests/unit/test_spaces.cpp [hunk 4]
```diff
@@ -115,25 +135,21 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
 
     dist_func_t<float> arch_opt_func;
     float baseline = FP32_L2Sqr(v, v2, dim);
-    switch (optimization) {
-    case ARCH_OPT_AVX512_BW_VBMI2:
-    case ARCH_OPT_AVX512_F:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F);
+    if (optimization.features.avx512f) {
+        arch_opt_func = L2_FP32_GetDistFunc(dim, optimization);
         ASSERT_EQ(baseline, arch_opt_func(v, v2, dim)) << "AVX512 with dim " << dim;
```
Comment (meiravgri) [inline tests/unit/test_spaces.cpp:140]:
```suggestion
        ASSERT_EQ(baseline, arch_opt_func(v, v2, dim)) << "AVX512 with dim " << dim;
        ASSERT_EQ(arch_opt_func, Choose_FP32_IP_implementation_AVX512(dim)) << "upexpected function for AVX512 with dim " << dim;
```

File: tests/unit/test_spaces.cpp [hunk 5]
```diff
@@ -115,25 +135,21 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
 
     dist_func_t<float> arch_opt_func;
     float baseline = FP32_L2Sqr(v, v2, dim);
-    switch (optimization) {
-    case ARCH_OPT_AVX512_BW_VBMI2:
-    case ARCH_OPT_AVX512_F:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F);
+    if (optimization.features.avx512f) {
```
Comment (meiravgri) [inline tests/unit/test_spaces.cpp:138]: either check also the compiler flags or add a comment that we assume that the compiler supports the highest arch optimization

File: tests/unit/CMakeLists.txt [hunk 6]
```diff
@@ -20,6 +20,8 @@ set(CMAKE_CXX_FLAGS_RELEASE "-O3")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CLANG_SAN_FLAGS} ${LLVM_CXX_FLAGS}")
 set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} ${LLVM_LD_FLAGS}")
 
+include(CheckCXXCompilerFlag)
```
Comment (meiravgri) [inline tests/unit/CMakeLists.txt:23]: needed?

File: tests/benchmark/spaces_benchmarks/bm_spaces.h [hunk 7]
```diff
@@ -13,19 +13,25 @@
...
 
 // Defining the generic benchmark flow: if there is support for the optimization, benchmark the
 // function.
-#define BENCHMARK_DISTANCE_F(type_prefix, arch, metric, bm_name)                                   \
+#define BENCHMARK_DISTANCE_F(type_prefix, arch, metric, bm_name, opt_supported)                    \
     BENCHMARK_DEFINE_F(BM_VecSimSpaces, type_prefix##_##arch##_##metric##_##bm_name)               \
     (benchmark::State & st) {                                                                      \
-        if (opt < spaces::ARCH_OPT_##arch) {                                                       \
+        if (!opt_supported) {                                                                      \
             st.SkipWithError("This benchmark requires " #arch ", which is not available");         \
```
Comment (meiravgri) [inline tests/benchmark/spaces_benchmarks/bm_spaces.h:31]: for consistency with "arch" param
```suggestion
#define BENCHMARK_DISTANCE_F(type_prefix, arch, metric, bm_name, arch_supported)                    \
    BENCHMARK_DEFINE_F(BM_VecSimSpaces, type_prefix##_##arch##_##metric##_##bm_name)               \
    (benchmark::State & st) {                                                                      \
        if (!arch_supported) {                                                                      \
            st.SkipWithError("This benchmark requires " #arch ", which is not available");         \
```

File: tests/benchmark/spaces_benchmarks/bm_spaces_class.h [hunk 8]
```diff
@@ -15,20 +15,20 @@ class BM_VecSimSpaces : public benchmark::Fixture {
...
 public:
     BM_VecSimSpaces();
     ~BM_VecSimSpaces() = default;
 
     void SetUp(const ::benchmark::State &state);
     void TearDown(const ::benchmark::State &state);
+
+    // Specific architecture optimization flags that are supported on this machine,
+    // to be initialized in every executable that is running this benchmarks.
+    static spaces::Arch_Optimization opt;
```
Comment (meiravgri) [inline tests/benchmark/spaces_benchmarks/bm_spaces_class.h:28]: is this is use?

**Review by meiravgri — COMMENTED** [review]
cool!
to avoid repetition i only added one comment for things that are relevant for more than one type/space/arch

File: tests/unit/test_spaces.cpp [hunk 9]
```diff
@@ -115,26 +135,24 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
...
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F);
+#ifdef OPT_AVX512F
+    if (optimization.features.avx512f) {
+        arch_opt_func = Choose_FP32_L2_implementation_AVX512(dim);
         ASSERT_EQ(baseline, arch_opt_func(v, v2, dim)) << "AVX512 with dim " << dim;
-    case ARCH_OPT_AVX2:
-    case ARCH_OPT_AVX:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX);
+    }
+#endif
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_spaces.cpp]: I added this check for the test, but I don't think we need to check the case where arch support is higher than the compiler's (at least not in the scope of this PR), we didn't had it before also.

File: tests/unit/test_spaces.cpp [hunk 10]
```diff
@@ -115,26 +135,24 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
...
+        arch_opt_func = Choose_FP32_L2_implementation_SSE(dim);
         ASSERT_EQ(baseline, arch_opt_func(v, v2, dim)) << "SSE with dim " << dim;
-    case ARCH_OPT_NONE:
-        arch_opt_func = L2_FP32_GetDistFunc(dim, ARCH_OPT_NONE);
-        ASSERT_EQ(FP32_L2Sqr, arch_opt_func);
-        break;
-    default:
-        FAIL();
     }
+#endif
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_spaces.cpp]: added check

File: tests/unit/test_spaces.cpp (hunk 2 above)
Comment (alonre24) [reply tests/unit/test_spaces.cpp]: WIth the new validation for choosing the right function - yes

File: tests/benchmark/spaces_benchmarks/bm_spaces_class.h (hunk 8 above)
Comment (alonre24) [reply tests/benchmark/spaces_benchmarks/bm_spaces_class.h]: It has slightly changed due to the removal of `Arch_Optimization`, but yes - for storing the optimizations in the class and testing against it (as comment says)

Comment (meiravgri) [reply tests/benchmark/spaces_benchmarks/bm_spaces_class.h]: I mean that it is only used once. you don't really need to store it in the class,

File: tests/unit/test_spaces.cpp (hunk 3 above)
Comment (meiravgri) [reply tests/unit/test_spaces.cpp:143]: wasn't fixed 😄

File: tests/unit/test_spaces.cpp [hunk 11]
```diff
@@ -135,28 +135,37 @@ TEST_P(FP32SpacesOptimizationTest, FP32L2SqrTest) {
 
     dist_func_t<float> arch_opt_func;
     float baseline = FP32_L2Sqr(v, v2, dim);
-#ifdef OPT_AVX512F
-    if (optimization.features.avx512f) {
-        arch_opt_func = Choose_FP32_L2_implementation_AVX512(dim);
+    if (optimization.avx512f) {
+        arch_opt_func = L2_FP32_GetDistFunc(dim, &optimization);
+        ASSERT_EQ(arch_opt_func, Choose_FP32_L2_implementation_AVX512(dim))
+            << "Unexpected distance function chosen for dim " << dim;
```
Comment (meiravgri) [inline tests/unit/test_spaces.cpp:141]: Why not to mention for which arch we failed?

File: src/VecSim/spaces/spaces.cpp [hunk 12]
```diff
@@ -13,30 +13,26 @@ namespace spaces {
...
 
 void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<double> *out_func,
                  unsigned char *alignment) {
 
-    static const Arch_Optimization arch_opt = getArchitectureOptimization();
-
     if (metric == VecSimMetric_Cosine || metric == VecSimMetric_IP) {
 
-        *out_func = IP_FP64_GetDistFunc(dim, arch_opt, alignment);
+        *out_func = IP_FP64_GetDistFunc(dim, nullptr, alignment);
```
Comment (meiravgri) [inline src/VecSim/spaces/spaces.cpp:31]: why does codecov claim that its not covered? 🤔

File: tests/unit/test_spaces.cpp (hunk 11 above)
Comment (alonre24) [reply tests/unit/test_spaces.cpp:141]: It can be inferred from the line number

File: src/VecSim/spaces/spaces.cpp (hunk 12 above)
Comment (alonre24) [reply src/VecSim/spaces/spaces.cpp:31]: Not sure, we are running FP64 tests on coverage, right?

File: tests/benchmark/spaces_benchmarks/bm_spaces_class.h (hunk 8 above)
Comment (alonre24) [reply tests/benchmark/spaces_benchmarks/bm_spaces_class.h]: Not sure what do you mean, let's discuss it

**Review by meiravgri — APPROVED** [review]

**Pull request merged**